app/ui/legacy/src/main/java/com/fsck/k9/ui/messagelist/SwipeResourceProvider.kt

```diff
@@ -12,15 +12,24 @@ import com.fsck.k9.ui.resolveColorAttribute
 class SwipeResourceProvider(val theme: Theme) {
     val iconTint = theme.resolveColorAttribute(R.attr.messageListSwipeIconTint)
 
-    private val selectIcon = theme.loadDrawable(Icons.Outlined.CheckCircle)
-    private val markAsReadIcon = theme.loadDrawable(Icons.Outlined.MarkEmailRead)
-    private val markAsUnreadIcon = theme.loadDrawable(Icons.Outlined.MarkEmailUnread)
-    private val addStarIcon = theme.loadDrawable(Icons.Filled.Star)
-    private val removeStarIcon = theme.loadDrawable(Icons.Outlined.Star)
-    private val archiveIcon = theme.loadDrawable(Icons.Outlined.Archive)
-    private val deleteIcon = theme.loadDrawable(Icons.Outlined.Delete)
-    private val spamIcon = theme.loadDrawable(Icons.Outlined.Report)
-    private val moveIcon = theme.loadDrawable(Icons.Outlined.DriveFileMove)
+    private val selectIcon: Drawable
+        get() = theme.loadDrawable(Icons.Outlined.CheckCircle)
```

Comment:
This changes the behavior. Due to now calling `mutate()` a new `Drawable` instance is created each time one of these properties is accessed (each time the swipe handler is called).

We can keep the feature of not eagerly loading, but only loading the drawable once by using `by lazy`.

```suggestion
    private val selectIcon by lazy(mode = LazyThreadSafetyMode.NONE) { theme.loadDrawable(Icons.Outlined.CheckCircle) }
```

Reply:
But is this a problem?

Reply:
I think the whole `SwipeResourceProvider` is flawed keeping around instances of resources. What problem is this solving in the first place. Resources are designed to be quickly allocated and released. Why this caching layer?

Reply:
The code will constantly create new instances while the user drags a message and while the view is animated back into its original position. Allocations in a hot code path is something that we should avoid and in this case easily can.

Reply:
I'm always interested to learn more. Especially as I worked with even bigger images allocated on the fly and wondering is a theoretical problem or really an issue. Had a look at the memory allocations and cpu usage and can't spot an issue versus a version with the mentioned changes. When trying to do a lot of swipes I noticed that allocations are not even close to be a problem. But the whole interatcion is super slow leading to drops an stuttering regardless if using the old, my or the suggested change version.

Reply:
After reading reading `MessageListSwipeCallback` I understand that there are some design issues in the way it allocated and accesses resources and draws the swipe action. I think it needs to be optimized to work properly. For example it uses `findViewById` in the `onChildDraw`, which is very slow.

But honestly a full rewrite is out of scope for this. I'll just update the `SwipeResourceProvider` implementation a bit.

Reply:
I added a print statement after this line:
https://github.com/thunderbird/thunderbird-android/blob/685dd91218682021dbb8c2871f131ac124221182/app/ui/legacy/src/main/java/com/fsck/k9/ui/messagelist/MessageListSwipeCallback.kt#L191

The output without the change from this PR applied:
```text
2024-05-28 11:26:58.346 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.355 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.361 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.369 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.378 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.386 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.395 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.403 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.411 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
2024-05-28 11:26:58.419 30282-30282 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@c4f846f
```

`android.graphics.drawable.VectorDrawable` inherits `toString()` and `hashCode()` from `Object`. So we can be reasonably sure that all of these are the same object, not different objects that all produce the same `hashCode()` value (`c4f846f`).

The output from this PR branch:
```text
2024-05-28 11:09:28.902 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@fb824f3
2024-05-28 11:09:28.908 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@ed13f6b
2024-05-28 11:09:28.917 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@82ccac8
2024-05-28 11:09:28.924 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@f8b7761
2024-05-28 11:09:28.935 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@2c79286
2024-05-28 11:09:28.942 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@7426747
2024-05-28 11:09:28.950 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@aa7cb74
2024-05-28 11:09:28.958 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@927289d
2024-05-28 11:09:28.966 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@a7a5112
2024-05-28 11:09:28.975 28801-28801 System.out              com.fsck.k9.debug                    I  android.graphics.drawable.VectorDrawable@b430e3
```

We can see that a) `drawLayout()` is invoked a lot and b) each time a new `VectorDrawable` instance is used. This might not be noticeable when the swipe handling is all the app is doing. However, when the app is running other things in the background (e.g. running a sync) there might be enough stress on the garbage collector that this could introduce jank.

With the suggested change the output goes back to being the same in each line.